### PR TITLE
[wasm] Run `Wasm.Build.Tests` against workloads

### PR DESCRIPTION
**Pull request by radical**
Currently we run `Wasm.Build.Tests` on helix, by setting various paths to reference files from `artifacts`. This PR adds another mode, we *try* to test against workload packs.

For this:

1. we need the nugets for the various workload packs, and the manifest
2. manifest id, path to those nugets
3. and a SDK version to test

Steps:

1. The specified SDK is installed in `artifacts/bin/dotnet-workload`
2. the specified manifest is installed from the built nugets, and any `depends-on` manifests in the json.
3. `dotnet workload install` is used to install the packs, with the built nugets path being used as one the nuget sourcs

Once this setup is done, then the tests are run in an environment such that they use `dotnet` from above, and try to resolve packs from there.

- We still want to test without packs, the case of using `EMSDK_PATH` directly, for example, in the library tests. So, we now run `Wasm.Build.Tests` for Workloads, *and* the regular `EMSDK_PATH`.

- Two workload specific tests were added in a recent PR (https://github.com/dotnet/runtime/pull/54936)
- simple test for building blazorwasm template project with AOT
- a test to validate that the files in `UnixFilePermissions.xml` for the packs actually exist on disk

.. and these are enabled here.

- To use this, the project sets `TestUsingWorkloads=true`, and specifies the workload with:

```xml
    <WorkloadIdForTesting Include="wasm-tools"
                           Name="microsoft.net.workload.mono.toolchain"
                           ManifestName="Microsoft.NET.Workload.Mono.ToolChain"
                           Version="$(PackageVersion)"
                           VersionBand="$(SdkBandVersion)" />
```

Notes:
- This also fixes using the correct runtime pack based on the version in the manifest json.
- The sdk version is specified with `<SdkVersionForWorkloadTesting>6.0.100-preview.7.21326.4</SdkVersionForWorkloadTesting>` in `eng/Versions.props`
- Known issues: I'm explicitly invoking `Pack` on the aotcross pkgproj

TODO:
- Run the blazorwasm test with playwright
- Add sdk-with-no-packs case also

### CODE REVIEW

**Comment by radical** [comment]
cc @lewing @steveisok RFC!

File: src/mono/nuget/Microsoft.NET.Runtime.WebAssembly.Sdk/Sdk/AutoImport.props.in [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<Project>
+  <PropertyGroup>
+    <WasmNativeWorkload>true</WasmNativeWorkload>
+    <UseMonoRuntime>true</UseMonoRuntime>
```
[the remark this answers is not in the document]
Comment (lewing) [reply src/mono/nuget/Microsoft.NET.Runtime.WebAssembly.Sdk/Sdk/AutoImport.props.in:4]: But the right place to still all these is in Sdk.props I think

File: src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in [hunk 2]
```diff
@@ -43,4 +48,21 @@
...
+
+    <Target Name="UpdateFrameworkReferences" BeforeTargets="ProcessFrameworkReferences">
+      <ItemGroup>
+        <KnownRuntimePack
+            Include="Microsoft.NETCore.App"
+            TargetFramework="net6.0"
+            LatestRuntimeFrameworkVersion="${PackageVersion}"
+            RuntimePackRuntimeIdentifiers="linux-x64;osx-x64;browser-wasm"
+            RuntimePackNamePatterns="Microsoft.NETCore.App.Runtime.Mono.**RID**"
+            RuntimePackLabels="BrowserWorkload"
```
[the remark this answers is not in the document]
Comment (steveisok) [reply src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in:60]: I also think the label might be important for the mono runtime packs. 

@akoeplinger would know for sure.

[the remark this answers is not in the document]
Comment (radical) [reply src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in:60]: So, label should be `MonoWorkload`?

Also, is this correct/enough - `RuntimePackRuntimeIdentifiers="linux-x64;osx-x64;browser-wasm`?

File: src/mono/nuget/Microsoft.NET.Runtime.WebAssembly.Sdk/Sdk/AutoImport.props.in [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+    <RuntimePackInWorkloadVersion>${PackageVersion}</RuntimePackInWorkloadVersion>
```
[the remark this answers is not in the document]
Comment (radical) [reply src/mono/nuget/Microsoft.NET.Runtime.WebAssembly.Sdk/Sdk/AutoImport.props.in:5]: How do I access this version during the build?

[the remark this answers is not in the document]
Comment (lewing) [reply src/mono/nuget/Microsoft.NET.Runtime.WebAssembly.Sdk/Sdk/AutoImport.props.in:5]: which build?

File: src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in (hunk 2 above)
[the remark this answers is not in the document]
Comment (lewing) [reply src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in:60]: > I also think the label might be important for the mono runtime packs.
> 
> @akoeplinger would know for sure.

The label is definitely important,  the details are important here

[the remark this answers is not in the document]
Comment (steveisok) [reply src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in:60]: I think we should override the existing mono `KnownRuntimePack` by first removing it and then replacing with a copy.  The only difference should be the `LatestRuntimeFrameworkVersion`, which should match what's in the workload.

The original:
https://github.com/dotnet/installer/blob/823ff33acf9c67374d4b1211850b22d3db7f020d/src/redist/targets/GenerateBundledVersions.targets#L314-L321

[the remark this answers is not in the document]
Comment (steveisok) [reply src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in:60]: I also wonder if we need `<Target Name="UpdateFrameworkReferences" BeforeTargets="ProcessFrameworkReferences">`?  We may just be able to condition an ItemGroup and do all the magic there.

[the remark this answers is not in the document]
Comment (steveisok) [reply src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in:60]: Perhaps something like:

```
<PropertyGroup Condition="'$(TargetPlatformIdentifier)' == 'android' or '$(TargetPlatformIdentifier)' == 'ios' or '$(TargetPlatformIdentifier)' == 'maccatalyst'
                            or '$(TargetPlatformIdentifier)' == 'tvos' or ($(RuntimeIdentifier)' == 'browser-wasm' and '$(UsingBrowserRuntimeWorkload)' == 'true')">
    <_MonoWorkloadTargetsMobile>true</_MonoWorkloadTargetsMobile>
    <_MonoWorkloadRuntimePackPackageVersion>${PackageVersion}</_MonoWorkloadRuntimePackPackageVersion>
  </PropertyGroup>
  
  <ItemGroup Condition="'$(_MonoWorkloadTargetsMobile)' == 'true'">
    <KnownRuntimePack Remove="Microsoft.NETCore.App" />
    <KnownRuntimePack Include="Microsoft.NETCore.App"
                      TargetFramework="net6.0"
                      RuntimeFrameworkName="Microsoft.NETCore.App"
                      LatestRuntimeFrameworkVersion="$(_MonoWorkloadRuntimePackPackageVersion)"
                      RuntimePackNamePatterns="Microsoft.NETCore.App.Runtime.Mono.**RID**"
                      RuntimePackRuntimeIdentifiers="@(MonoRuntimePackRids, '%3B')"
                      RuntimePackLabels="Mono"
                      />
  </ItemGroup>
  ```

File: src/mono/nuget/Microsoft.NET.Runtime.WebAssembly.Sdk/Sdk/AutoImport.props.in (hunk 3 above)
[the remark this answers is not in the document]
Comment (radical) [reply src/mono/nuget/Microsoft.NET.Runtime.WebAssembly.Sdk/Sdk/AutoImport.props.in:5]: A user project build, or tests. I need the version for the tests, and right now it is being used to figure out the path for the runtime pack too. Though as you had suggested, we could move to using `%(ResolvedRuntimePack.PackageDirectory)`, but that needs some changes.

File: src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in (hunk 2 above)
[the remark this answers is not in the document]
Comment (radical) [reply src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in]: That seems to work fine

**Comment by radical** [comment]
Just doing some final tidbits, but this should be good for reviewing.

File: src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/Sdk.props [hunk 4]
```diff
@@ -1,4 +1,8 @@
 <Project>
+  <PropertyGroup>
```
Comment (steveisok) [inline src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/Sdk.props:2]: As I understand, coreclr is going to be the default target for xamarin macos.  So, blanket setting this property is not going to work.  

There is the ability to switch runtimes, but I'm not sure if that is going to be a supported feature or not.

Comment (radical) [reply src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/Sdk.props:2]: Hmm so, RID=browser-wasm for now?

File: src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/Sdk.props [hunk 5]
```diff
@@ -1,4 +1,8 @@
...
+    <UseMonoRuntime Condition="'$(RuntimeIdentifier)' == 'Browser-wasm'">true</UseMonoRuntime>
```
Comment (akoeplinger) [inline src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/Sdk.props:3]: I think/hope this is no longer needed after my #55258

Comment (radical) [reply src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/Sdk.props:3]: Your PR changes the InTree, and LocalBuild targets, which don't get used for the workloads.

**Comment by radical** [comment]
`Windows x86` failure is https://github.com/dotnet/runtime/issues/55611

File: src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/Sdk.props (hunk 5 above)
Comment (radical) [reply src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/Sdk.props:3]: Follow up PR? :) And if we are moving it to wasm sdk, then might as well move it to wasm props, then it won't be needed in InTree/LocalBuild either.

**Review by lewing — APPROVED** [review]
Discussed some follow up with @radical but since we're already have a pending pr building off this the changes can happen there

**Pull request merged**